exercises/sat/descr.md

```diff
@@ -0,0 +1,616 @@
+created so far*; `new_clause c` adds the clause `c` to a growing conjunction
+of *all clauses declared so far*.
+
+Using these facilities, you are expected to implement the function `cnf`,
+which receives a formula `f` and must construct an equisatisfiable formula
+in conjunctive normal form. This construction is carried out via suitable
+calls to `new_var` and `new_clause`. No result is returned.
+
+**Question 4.** Implement the function `cnf`.
+
+*Hint.* We suggest writing two auxiliary functions, whose types are as
+follows:
+
+```
+  val decompose: bool -> formula -> clause -> unit
+  val clause: bool -> formula -> clause -> clause
+```
+
+(These functions should be mutually recursive.) These functions expect
+the same parameters, namely a Boolean polarity `s`, a formula `f`, and
+a clause `c`, which together represent the logical formula `s.f \/ c`.
+(We write `s.f` for `f` if `s` is `true` and `~f` if `s` is `false`.)
+
+The function `decompose` is expected to convert `s.f \/ c` to CNF, via
+suitable calls to `new_var` and `new_clause`. It returns no result.
+
+The function `clause` is expected to convert `s.f \/ c` to a clause,
+which it returns. It is also allowed to call `new_var` and `new_clause`
+so as to construct other clauses on the side.
+
+*Hint.* The easy case in `decompose` is when `s.f` is a conjunction, because
+it then suffices to handle each conjunct independently. The easy case in
+`clause` is when `s.f` is a disjunction, because it then suffices to deal with
+each disjunct in turn and combine the results into a single clause.
+
+## Recreation: Intersection of Two Sorted Lists
+
+**Question 5.** Implement a function `intersect`
+of type `int list -> int list -> bool`
+such that `intersect xs ys` is `true`
+if and only if the two *sorted* lists
+`xs` and `ys` have a common element.
+
+## Representing Clauses
+
+We would now like to choose a representation of clauses in memory and
+implement the functions `empty`, `cons`, `new_clause` and `new_var`
+that the functor `CNF` (above) requires.
+
+We adopt the convention that a literal is represented as a pair of a Boolean
+polarity and a variable:
+
+```
+  type literal = bool * var
+```
+
+We adopt the convention that a clause is represented as a list of literals:
+
+```
+  type clause = literal list
+```
+
+**Question A.** (Ungraded.) Implement the constant `empty` and the constructor
+function `cons`.
+
+We adopt the convention that clauses are numbered
+and are stored in an infinite array `clauses`.
+Infinite arrays are described at the very end of this exercise.
+
+**Question B.** (Ungraded.) Implement the function `new_clause`, which records a
+new clause, and the function `count_clauses`, which indicates how many clauses
+have been recorded so far. Note that a clause that contains both the literal
+`x` and the opposite literal `~x` is unsatisfiable. In such a situation, the
+function `new_clause` must raise the exception `UNSAT`. *Hint:* to detect this
```

Comment:
I suggest moving *Hint* to the beginning of the next line. I want to be able to automatically detect these paragraphs in the future to hide them unless the student explicitly asks for help.

```suggestion
function `new_clause` must raise the exception `UNSAT`. 

*Hint:* to detect this
```